core/include/scipp/core/element_array.h

```diff
@@ -77,7 +81,7 @@ template <class T> class element_array {
   }
 
   element_array(const element_array &other)
-      : element_array(other.data(), other.data() + other.size()) {}
+      : element_array(other.begin(), other.end(), other.size()) {}
```

Comment:
I wouldn't mix the constructor for normal iterators let's `std::vector::iterator` with the case of copy constructor for `element_array`. It is better to introduce the private fabric function let say `make_element_array(const element_array & other)` and call it. Then to avoid usage of wrong constructor with iterators lets say this shouldn't compile:
```c++
element_array arr;
element_array arr1(arr.begin(), arr.end());
``` 
I would change signature to:
```
template <class Iter>
std::enable_if_t<std::is_assignable<T &, Iter::value_type)>{}, int> = 0>
element_array(Iter first, Iter last);
```

Reply:
> It is better to introduce the private fabric function let say make_element_array(const element_array & other) and call it

I don't understand where that function would be called: in a new copy constructor, or whenever we want to do 

```
element_array arr;
element_array arr1(arr.begin(), arr.end());
```

we would instead have to do

```
element_array arr;
element_array arr1 = arr.make_element_array()
```

Reply:
> > It is better to introduce the private fabric function let say make_element_array(const element_array & other) and call it
> 
> I don't understand where that function would be called: in a new copy constructor, or whenever we want to do
> 
> ```
> element_array arr;
> element_array arr1(arr.begin(), arr.end());
> ```
> 
> we would instead have to do
> 
> ```
> element_array arr;
> element_array arr1 = arr.make_element_array(arr)
> ```

But now you just expose  the broken API, I'm a user and I can write like this:
```c++
element_array arr;
element_array arr1(arr.begin(), arr.end());
```
and have no compilation error, but the wrong result.
I'm suggesting:
```c++
private:
  static element_array from_other(const element_array& arr) {
    if (arr.size() == -1)
      return element_array();
    if (arr.size() = 0)
       return element_array(0);
    else ....
  }
public:
  element_array(const element_array& other) : element_array(from_other(other));
```